src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs

```diff
@@ -1597,17 +1598,33 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
                         // different from the local variable `type`, which has the node's type substituted for the current container.
                         var cacheVariableType = containerAsFrame.TypeMap.SubstituteType(node.Type).Type;
 
-                        var cacheVariableName = GeneratedNames.MakeLambdaCacheFieldName(
-                            // If we are generating the field into a display class created exclusively for the lambda the lambdaOrdinal itself is unique already, 
-                            // no need to include the top-level method ordinal in the field name.
-                            (closureKind == ClosureKind.General) ? -1 : topLevelMethodId.Ordinal,
-                            topLevelMethodId.Generation,
-                            lambdaId.Ordinal,
-                            lambdaId.Generation);
-
-                        var cacheField = new SynthesizedLambdaCacheFieldSymbol(translatedLambdaContainer, cacheVariableType, cacheVariableName, _topLevelMethod, isReadOnly: false, isStatic: closureKind == ClosureKind.Singleton);
-                        CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(translatedLambdaContainer, cacheField);
-                        cache = F.Field(receiver, cacheField.AsMember(constructedFrame)); //NOTE: the field was added to the unconstructed frame type.
+                        // Also we want to substitute the type according to the type map for the method that uses the delegate.
+                        // Such substitution handles cases when the delegate type uses a type parameter from the method
+                        // and the parameter is not captured by the container.
+                        if (TryGetGenericMethodTypeMap(referencedMethod.ConstructedFrom, out var typeMap))
+                        {
+                            cacheVariableType = typeMap.SubstituteType(cacheVariableType).Type;
```

Comment:
>cacheVariableType = typeMap.SubstituteType(cacheVariableType).Type; [](start = 28, length = 67)

Why is this needed? it doesn't look like we did this before.

Reply:
After closer investigation, I am convinced that we are approaching this in a somewhat wrong way. After debugging the test, I realized that ```referencedMethod` is the lambda that is being converted to the delegate, it doesn't really provide a context for the delegate type, the context comes from an enclosing method and enclosing types. In terms of the comment below: "We cannot reference type parameters from the parent method outside the method", the ```referencedMethod``` is not a parent method.
Even though the check below appears to provide expected value in our scenarios, this is not the question we should be asking. The better (and simpler) question to ask is whether the ```cacheVariableType``` as calculated above refers to any method type parameters (I mean from any method) after the substitution. If that is the case, we cannot use it for a field in any type. I don't think we have a ready to use helper to perform a check like that, but it should be pretty straight forward to add one next to ```ContainsTypeParameter``` used below following the same implementation strategy.

---
In reply to: [447782927](https://github.com/dotnet/roslyn/pull/44939#discussion_r447782927) [](ancestors = 447782927)

Reply:
> The better (and simpler) question to ask is whether the cacheVariableType as calculated above refers to any method type parameters (I mean from any method) after the substitution. 

Consider the following code. In this case the lambda refers to another method type parameter but there is nothing wrong to cache the delegate because the type parameter has larger scope than the cache.
```
void GenericMethod<T>(T value)
{
    for (int i = 0; i < 1; i++)
    {
        DoStuff<T>(() => value);
    }
}

void DoStuff<T>(Func<T> lambda)
{
    Console.WriteLine(lambda());
}
```

I'm totally ok to implement the fix in any way you think is the best. But I'm worried with this kind of fix, because the fix you offer eliminates the cache in that scenario hence may slow down some code that already works fine.

Reply:
> Consider the following code. In this case, the lambda refers to another method type parameter, but there is nothing wrong to cache the delegate because the type parameter has a larger scope than the cache.

The concept of a "larger scope" for a type parameter probably needs a clear definition. And I think the suggested fix can esily be adjusted to ignore type parameters like that. However, if we are talking about method type parameters that are becoming type type parameters, then, perhaps, the "after the substitution" is the important part in the suggestion. Is it possible that "after the substitution" they become type type parameters and, therefore, are no longer treated as "bad" by the suggested fix? It has been a long time since I looked at this PR, so it is quite possible I am missing something.

Reply:
I've implemented the fix you offered. Please, take a look.

What about the case I mentioned, you fix handles it perfectly fine. The lambda is already "moved" into a display class, so from the perspective of the fix the type argument of `GenericMethod` is actually a type argument of the display class.
I've added a unit test to verify that such a case works fine after the fix.